File: src/events/engine.ts
/**
 * 이벤트 엔진
 * 중앙 이벤트 버스 및 이벤트 처리를 담당합니다.
 */

import EventEmitter from 'eventemitter3';
import {
  BaseEvent,
  EventHandler,
  EventFilter,
  EventTransformer,
  EventSubscriptionOptions,
  EventPublishOptions,
  EventBatch,
  EventStatistics,
  EventCategory,
  EventSeverity,
} from './types/index.js';
// queueManager는 나중에 동적으로 임포트

/**
 * 이벤트 구독자 정보
 */
interface EventSubscriber<T extends BaseEvent = BaseEvent> {
  id: string;
  pattern: string | RegExp;
  handler: EventHandler<T>;
  options: EventSubscriptionOptions;
  createdAt: Date;
}

/**
 * 이벤트 통계 데이터
 */
interface EventStats {
  totalEvents: number;
  eventsByCategory: Map<EventCategory, number>;
  eventsBySeverity: Map<EventSeverity, number>;
  eventsPerHour: number[];
  lastEventTime: Date | undefined;
}

/**
 * 이벤트 엔진 클래스
 */
export class EventEngine extends EventEmitter {
  private subscribers: Map<string, EventSubscriber[]> = new Map();
  private eventQueue: BaseEvent[] = [];
  private stats: EventStats;
  private transformers: Map<string, EventTransformer[]> = new Map();
  private globalFilters: EventFilter[] = [];
  private useQueueManager: boolean = true;
  private queueManager: any = null;

  constructor() {
    super();

    // 통계 초기화
    this.stats = {
      totalEvents: 0,
      eventsByCategory: new Map(),
      eventsBySeverity: new Map(),
      eventsPerHour: Array(24).fill(0),
      lastEventTime: undefined as Date | undefined,
    };

    // 카테고리별 통계 초기화
    Object.values(EventCategory).forEach((category) => {
      this.stats.eventsByCategory.set(category, 0);
    });

    // 심각도별 통계 초기화
    Object.values(EventSeverity).forEach((severity) => {
      this.stats.eventsBySeverity.set(severity, 0);
    });

    // QueueManager 이벤트 처리기 설정
    if (this.useQueueManager) {
      this.setupQueueManagerIntegration();
    }
  }

  /**
   * 이벤트 구독
   */
  subscribe<T extends BaseEvent = BaseEvent>(
    pattern: string | RegExp,
    handler: EventHandler<T>,
    options: EventSubscriptionOptions = {},
  ): string {
    const subscriberId = this.generateSubscriberId();
    const subscriber: EventSubscriber<T> = {
      id: subscriberId,
      pattern,
      handler: handler as EventHandler,
      options,
      createdAt: new Date(),
    };

    const key = pattern instanceof RegExp ? pattern.source : pattern;
    const subscribers = this.subscribers.get(key) || [];

    // 우선순위에 따라 정렬하여 추가
    const priority = options.priority || 0;
    const insertIndex = subscribers.findIndex((s) => (s.options.priority || 0) < priority);

    if (insertIndex === -1) {
      subscribers.push(subscriber as EventSubscriber);
    } else {
      subscribers.splice(insertIndex, 0, subscriber as EventSubscriber);
    }

    this.subscribers.set(key, subscribers);

    // EventEmitter3 이벤트 등록
    if (pattern instanceof RegExp) {
      // 정규식 패턴은 와일드카드로 처리
      this.on('*', (event: T) => {
        if (pattern.test(event.type)) {
          this.handleEvent(event, subscriber as EventSubscriber);
        }
      });
    } else {
      // 일반 문자열 패턴
      const eventHandler = (event: T) => this.handleEvent(event, subscriber as EventSubscriber);

      if (options.once) {
        this.once(pattern, eventHandler);
      } else {
        this.on(pattern, eventHandler);
      }
    }

    return subscriberId;
  }

  /**
   * 이벤트 구독 해제
   */
  unsubscribe(subscriberId: string): boolean {
    for (const [key, subscribers] of this.subscribers.entries()) {
      const index = subscribers.findIndex((s) => s.id === subscriberId);
      if (index !== -1) {
        subscribers.splice(index, 1);
        if (subscribers.length === 0) {
          this.subscribers.delete(key);
        }
        return true;
      }
    }
    return false;
  }

  /**
   * 이벤트 발행
   */
  async publish<T extends BaseEvent = BaseEvent>(
    event: T,
    options: EventPublishOptions = {},
  ): Promise<void> {
    // 통계 업데이트
    this.updateStatistics(event);

    // 전역 필터 적용
    if (!this.applyGlobalFilters(event)) {
      return;
    }

    // 변환기 적용
    const transformedEvent = await this.applyTransformers(event);

    // QueueManager 사용 시 큐로 라우팅
    if (this.useQueueManager && this.queueManager && (options as any).useQueue !== false) {
      const routed = await this.queueManager.routeEvent(transformedEvent);
      if (routed) {
        // 큐로 라우팅된 경우 직접 처리하지 않음
        this.emit('event:queued', transformedEvent);
        return;
      }
    }

    // 이벤트 큐에 추가
    this.eventQueue.push(transformedEvent);

    // EventEmitter3로 이벤트 발행 (processEvent에서 구독자 처리는 별도)
    this.emit(transformedEvent.type, transformedEvent);
    this.emit('*', transformedEvent);
    this.emit('event:published', transformedEvent);
  }

  /**
   * 배치 이벤트 발행
   */
  async publishBatch<T extends BaseEvent = BaseEvent>(
    batch: EventBatch<T>,
    options: EventPublishOptions = {},
  ): Promise<void> {
    for (const event of batch.events) {
      await this.publish(event, options);
    }
  }

  /**
   * 이벤트 변환기 등록
   */
  registerTransformer<T extends BaseEvent = BaseEvent, R extends BaseEvent = BaseEvent>(
    pattern: string | RegExp,
    transformer: EventTransformer<T, R>,
  ): void {
    const key = pattern instanceof RegExp ? pattern.source : pattern;
    const transformers = this.transformers.get(key) || [];
    transformers.push(transformer as unknown as EventTransformer<BaseEvent, BaseEvent>);
    this.transformers.set(key, transformers);
  }

  /**
   * 전역 필터 추가
   */
  addGlobalFilter(filter: EventFilter): void {
    this.globalFilters.push(filter);
  }

  /**
   * 전역 필터 제거
   */
  removeGlobalFilter(filter: EventFilter): boolean {
    const index = this.globalFilters.indexOf(filter);
    if (index !== -1) {
      this.globalFilters.splice(index, 1);
      return true;
    }
    return false;
  }

  /**
   * 통계 조회
   */
  getStatistics(): EventStatistics {
    return {
      totalEvents: this.stats.totalEvents,
      eventsByCategory: Object.fromEntries(this.stats.eventsByCategory) as Record<
        EventCategory,
        number
      >,
      eventsBySeverity: Object.fromEntries(this.stats.eventsBySeverity) as Record<
        EventSeverity,
        number
      >,
      eventsPerHour: this.calculateEventsPerHour(),
      ...(this.stats.lastEventTime && { lastEventTime: this.stats.lastEventTime }),
    };
  }

  /**
   * 이벤트 큐 크기 조회
   */
  getQueueSize(): number {
    return this.eventQueue.length;
  }

  /**
   * 구독자 수 조회
   */
  getSubscriberCount(): number {
    let count = 0;
    for (const subscribers of this.subscribers.values()) {
      count += subscribers.length;
    }
    return count;
  }

  /**
   * 이벤트 큐 비우기
   */
  clearQueue(): void {
    this.eventQueue = [];
  }

  /**
   * 모든 구독 해제
   */
  clearAllSubscriptions(): void {
    this.subscribers.clear();
    this.removeAllListeners();
  }

  /**
   * 이벤트 처리
   */
  private async handleEvent(event: BaseEvent, subscriber: EventSubscriber): Promise<void> {
    try {
      // 필터 적용
      if (subscriber.options.filter && !subscriber.options.filter(event)) {
        return;
      }

      // 핸들러 실행
      if (subscriber.options.async === false) {
        // 동기 실행
        subscriber.handler(event);
      } else {
        // 비동기 실행 (기본값)
        await subscriber.handler(event);
      }
    } catch (error) {
      // 에러 이벤트 발행
      const errorEvent: BaseEvent = {
        id: `error-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`,
        type: 'system:error',
        category: EventCategory.SYSTEM,
        timestamp: Date.now(),
        severity: EventSeverity.ERROR,
        source: 'EventEngine',
        data: {
          originalEvent: event,
          subscriberId: subscriber.id,
          error: error instanceof Error ? error.message : String(error),
          stack: error instanceof Error ? error.stack : (undefined as string | undefined),
        },
      };

      // 에러 이벤트는 직접 emit (무한 루프 방지)
      this.emit('system:error', errorEvent);
    }
  }

  /**
   * 통계 업데이트
   */
  private updateStatistics(event: BaseEvent): void {
    this.stats.totalEvents++;
    this.stats.lastEventTime = new Date();

    // 카테고리별 통계
    const categoryCount = this.stats.eventsByCategory.get(event.category) || 0;
    this.stats.eventsByCategory.set(event.category, categoryCount + 1);

    // 심각도별 통계
    const severityCount = this.stats.eventsBySeverity.get(event.severity) || 0;
    this.stats.eventsBySeverity.set(event.severity, severityCount + 1);

    // 시간대별 통계
    const hour = new Date().getHours();
    if (this.stats.eventsPerHour[hour] !== undefined) {
      this.stats.eventsPerHour[hour]++;
    }
  }

  /**
   * 전역 필터 적용
   */
  private applyGlobalFilters(event: BaseEvent): boolean {
    for (const filter of this.globalFilters) {
      if (!filter(event)) {
        return false;
      }
    }
    return true;
  }

  /**
   * 변환기 적용
   */
  private async applyTransformers(event: BaseEvent): Promise<BaseEvent> {
    let transformedEvent = event;

    // 정확한 타입 매칭 변환기
    const exactTransformers = this.transformers.get(event.type) || [];
    for (const transformer of exactTransformers) {
      transformedEvent = await transformer(transformedEvent);
    }

    // 정규식 패턴 변환기
    for (const [pattern, transformers] of this.transformers.entries()) {
      if (pattern.startsWith('/') && pattern.endsWith('/')) {
        const regex = new RegExp(pattern.slice(1, -1));
        if (regex.test(event.type)) {
          for (const transformer of transformers) {
            transformedEvent = await transformer(transformedEvent);
          }
        }
      }
    }

    return transformedEvent;
  }

  /**
   * 시간당 이벤트 수 계산
   */
  private calculateEventsPerHour(): number {
    const total = this.stats.eventsPerHour.reduce((sum, count) => sum + count, 0);
    return total / 24;
  }

  /**
   * 구독자 ID 생성
   */
  private generateSubscriberId(): string {
    return `sub-${Date.now()}-${Math.random().toString(36).substring(2, 9)}`;
  }

  /**
   * QueueManager 통합 설정
   */
  private async setupQueueManagerIntegration(): Promise<void> {
    try {
      // 동적 임포트로 순환 참조 해결
      const { queueManager } = await import('./queue-manager.js');
      this.queueManager = queueManager;
      
      // 기본 처리기 등록
      queueManager.registerProcessor('default', async (events) => {
        for (const event of events) {
          await this.processQueuedEvent(event);
        }
      });

      queueManager.registerProcessor('priority', async (events) => {
        for (const event of events) {
          await this.processQueuedEvent(event);
        }
      });

      queueManager.registerProcessor('batch', async (events) => {
        // 배치 처리 최적화
        await this.processBatchEvents(events);
      });

      // 실패한 이벤트 재처리
      queueManager.registerProcessor('failed', async (events) => {
        for (const event of events) {
          try {
            await this.processQueuedEvent(event);
          } catch (error) {
            console.error('Failed to reprocess event:', event.id, error);
          }
        }
      });

      // 큐 매니저 이벤트 리스닝
      queueManager.on('error', (error: Error, queueName?: string) => {
        console.error(`Queue error in ${queueName}:`, error);
        this.emit('queue:error', { error, queueName });
      });

      queueManager.on('stats:update', (stats: any) => {
        this.emit('queue:stats', stats);
      });
    } catch (error) {
      console.error('Failed to setup queue manager integration:', error);
      this.useQueueManager = false;
    }
  }

  /**
   * 이벤트와 매칭되는 구독자 찾기
   */
  private findMatchingSubscribers(event: BaseEvent): EventSubscriber[] {
    const subscribers: EventSubscriber[] = [];
    
    // 정확한 타입 매칭
    const exactMatch = this.subscribers.get(event.type);
    if (exactMatch) {
      subscribers.push(...exactMatch);
    }
    
    // 정규식 패턴 매칭
    for (const [pattern, subs] of this.subscribers.entries()) {
      if (pattern !== event.type) {
        try {
          const regex = new RegExp(pattern);
          if (regex.test(event.type)) {
            subscribers.push(...subs);
          }
        } catch {
          // 정규식이 아닌 경우 무시
        }
      }
    }
    
    return subscribers;
  }

  /**
   * 큐에서 가져온 이벤트 처리
   */
  private async processQueuedEvent(event: BaseEvent): Promise<void> {
    // 매칭되는 구독자 찾기
    const matchingSubscribers = this.findMatchingSubscribers(event);

    // 구독자에게 이벤트 전달
    for (const subscriber of matchingSubscribers) {
      try {
        // 구독자별 필터 적용
        if (subscriber.options.filter && !(await subscriber.options.filter(event))) {
          continue;
        }

        // 핸들러 실행
        await subscriber.handler(event);
      } catch (error) {
        console.error(`Error in event handler ${subscriber.id}:`, error);
        throw error; // 재시도를 위해 에러 전파
      }
    }
  }

  /**
   * 배치 이벤트 처리
   */
  private async processBatchEvents(events: BaseEvent[]): Promise<void> {
    // 타입별로 그룹화
    const eventsByType = new Map<string, BaseEvent[]>();
    
    for (const event of events) {
      const list = eventsByType.get(event.type) || [];
      list.push(event);
      eventsByType.set(event.type, list);
    }

    // 타입별로 배치 처리
    for (const [type, typeEvents] of eventsByType) {
      const subscribers = this.subscribers.get(type) || [];
      
      for (const subscriber of subscribers) {
        try {
          // 배치 핸들러가 있는 경우
          if ((subscriber.options as any).batchHandler) {
            await (subscriber.options as any).batchHandler(typeEvents);
          } else {
            // 개별 처리
            for (const event of typeEvents) {
              if (!subscriber.options.filter || await subscriber.options.filter(event)) {
                await subscriber.handler(event);
              }
            }
          }
        } catch (error) {
          console.error(`Error in batch handler ${subscriber.id}:`, error);
        }
      }
    }
  }

  /**
   * 큐 관리자 활성화/비활성화
   */
  async setUseQueueManager(enabled: boolean): Promise<void> {
    this.useQueueManager = enabled;
    if (enabled && !this.queueManager) {
      await this.setupQueueManagerIntegration();
    }
  }

  /**
   * 큐 통계 가져오기
   */
  getQueueStats(): Map<string, any> | null {
    if (this.useQueueManager && this.queueManager) {
      return this.queueManager.getAllStats();
    }
    return null;
  }

  /**
   * 큐 매니저 가져오기
   */
  getQueueManager() {
    return this.queueManager;
  }

  /**
   * 통계 가져오기
   */
  getStats() {
    return {
      totalEvents: this.stats.totalEvents,
      lastEventTime: this.stats.lastEventTime,
      eventsByCategory: Object.fromEntries(this.stats.eventsByCategory),
      eventsBySeverity: Object.fromEntries(this.stats.eventsBySeverity),
      eventsPerHour: this.stats.eventsPerHour,
      subscriberCount: this.subscribers.size,
      transformerCount: Array.from(this.transformers.values()).flat().length,
      globalFilterCount: this.globalFilters.length
    };
  }
}

// 싱글톤 인스턴스
export const eventEngine = new EventEngine();

File: python/src/devflow_monitor/events/engine.py
"""
Event Engine.

Central event bus and event processing system. Provides pub/sub
functionality with pattern matching, priority-based subscription,
and async event processing.
"""

from __future__ import annotations

import asyncio
import re
from collections import deque
from datetime import datetime
from typing import Any, Callable, Coroutine
from uuid import uuid4

from .types import (
    BaseEvent,
    EventBatch,
    EventCategory,
    EventFilter,
    EventPublishOptions,
    EventSeverity,
    EventStatistics,
    EventSubscriptionOptions,
)


# Type alias for event handlers
EventHandlerType = Callable[[BaseEvent], Coroutine[Any, Any, None] | None]


class EventSubscriber:
    """Event subscriber information."""

    def __init__(
        self,
        subscriber_id: str,
        pattern: str | re.Pattern[str],
        handler: EventHandlerType,
        options: EventSubscriptionOptions,
    ):
        """
        Initialize event subscriber.

        Args:
            subscriber_id: Unique subscriber identifier.
            pattern: Event type pattern (string or regex).
            handler: Event handler function.
            options: Subscription options.
        """
        self.id = subscriber_id
        self.pattern = pattern
        self.handler = handler
        self.options = options
        self.created_at = datetime.utcnow()

    def matches(self, event_type: str) -> bool:
        """
        Check if event type matches this subscriber's pattern.

        Args:
            event_type: Event type string to match.

        Returns:
            True if pattern matches, False otherwise.
        """
        if isinstance(self.pattern, re.Pattern):
            return bool(self.pattern.match(event_type))
        elif self.pattern == "*":
            return True
        else:
            return self.pattern == event_type


class EventStats:
    """Event statistics tracking."""

    def __init__(self):
        """Initialize event statistics."""
        self.total_events = 0
        self.events_by_category: dict[str, int] = {}
        self.events_by_severity: dict[str, int] = {}
        self.events_per_hour: list[int] = [0] * 24
        self.last_event_time: datetime | None = None

        # Initialize category counts
        for category in EventCategory:
            self.events_by_category[category.value] = 0

        # Initialize severity counts
        for severity in EventSeverity:
            self.events_by_severity[severity.value] = 0


class EventEngine:
    """
    Event Engine class.

    Central event bus for publishing and subscribing to events.
    Supports pattern matching, priority-based subscription,
    and async event processing.

    Attributes:
        max_history: Maximum number of events to keep in history.
    """

    _instance: EventEngine | None = None

    def __init__(self, max_history: int = 10000):
        """
        Initialize event engine.

        Args:
            max_history: Maximum number of events to keep in history.
        """
        self._subscribers: dict[str, list[EventSubscriber]] = {}
        self._event_history: deque[BaseEvent] = deque(maxlen=max_history)
        self._stats = EventStats()
        self._transformers: dict[str, list[Callable]] = {}
        self._global_filters: list[EventFilter] = []
        self._listeners: dict[str, list[EventHandlerType]] = {}
        self._queue_manager: Any = None
        self._use_queue_manager = False
        self._max_history = max_history

    @classmethod
    def get_instance(cls) -> EventEngine:
        """
        Get singleton instance of EventEngine.

        Returns:
            EventEngine singleton instance.
        """
        if cls._instance is None:
            cls._instance = EventEngine()
        return cls._instance

    @classmethod
    def reset_instance(cls) -> None:
        """Reset the singleton instance (useful for testing)."""
        cls._instance = None

    def subscribe(
        self,
        pattern: str | re.Pattern[str],
        handler: EventHandlerType,
        options: EventSubscriptionOptions | None = None,
    ) -> str:
        """
        Subscribe to events matching a pattern.

        Args:
            pattern: Event type pattern (string, regex, or "*" for all).
            handler: Event handler function.
            options: Subscription options.

        Returns:
            Subscription ID for later unsubscription.
        """
        if options is None:
            options = EventSubscriptionOptions()

        subscriber_id = self._generate_subscriber_id()
        subscriber = EventSubscriber(subscriber_id, pattern, handler, options)

        # Get pattern key for storage
        key = pattern.pattern if isinstance(pattern, re.Pattern) else pattern

        if key not in self._subscribers:
            self._subscribers[key] = []

        # Insert by priority (higher priority first)
        subscribers = self._subscribers[key]
        priority = options.priority
        insert_index = len(subscribers)

        for i, sub in enumerate(subscribers):
            if sub.options.priority < priority:
                insert_index = i
                break

        subscribers.insert(insert_index, subscriber)

        return subscriber_id

    def unsubscribe(self, subscriber_id: str) -> bool:
        """
        Unsubscribe from events.

        Args:
            subscriber_id: Subscription ID returned from subscribe().

        Returns:
            True if successfully unsubscribed, False otherwise.
        """
        for key, subscribers in list(self._subscribers.items()):
            for i, sub in enumerate(subscribers):
                if sub.id == subscriber_id:
                    subscribers.pop(i)
                    if not subscribers:
                        del self._subscribers[key]
                    return True
        return False

    def on(self, event_type: str, handler: EventHandlerType) -> str:
        """
        Register an event listener (convenience method).

        Args:
            event_type: Event type to listen for.
            handler: Event handler function.

        Returns:
            Subscription ID.
        """
        return self.subscribe(event_type, handler)

    def off(self, subscriber_id: str) -> bool:
        """
        Remove an event listener (convenience method).

        Args:
            subscriber_id: Subscription ID.

        Returns:
            True if removed, False otherwise.
        """
        return self.unsubscribe(subscriber_id)

    async def publish(
        self,
        event: BaseEvent,
        options: EventPublishOptions | None = None,
    ) -> None:
        """
        Publish an event to all matching subscribers.

        Args:
            event: Event to publish.
            options: Publish options.
        """
        if options is None:
            options = EventPublishOptions()

        # Update statistics
        self._update_statistics(event)

        # Apply global filters
        if not self._apply_global_filters(event):
            return

        # Apply transformers
        transformed_event = await self._apply_transformers(event)

        # Add to history
        self._event_history.append(transformed_event)

        # Route to queue manager if enabled
        if self._use_queue_manager and self._queue_manager and options.use_queue:
            routed = await self._queue_manager.route_event(transformed_event)
            if routed:
                await self._emit("event:queued", transformed_event)
                return

        # Process subscribers
        await self._process_event_subscribers(transformed_event)

        # Emit published event
        await self._emit("event:published", transformed_event)

    async def publish_batch(
        self,
        batch: EventBatch,
        options: EventPublishOptions | None = None,
    ) -> None:
        """
        Publish a batch of events.

        Args:
            batch: Event batch to publish.
            options: Publish options.
        """
        for event in batch.events:
            await self.publish(event, options)

    def register_transformer(
        self,
        pattern: str | re.Pattern[str],
        transformer: Callable[[BaseEvent], BaseEvent | Coroutine[Any, Any, BaseEvent]],
    ) -> None:
        """
        Register an event transformer.

        Args:
            pattern: Event type pattern.
            transformer: Transformer function.
        """
        key = pattern.pattern if isinstance(pattern, re.Pattern) else pattern
        if key not in self._transformers:
            self._transformers[key] = []
        self._transformers[key].append(transformer)

    def add_global_filter(self, filter_func: EventFilter) -> None:
        """
        Add a global event filter.

        Args:
            filter_func: Filter function that returns True to allow event.
        """
        self._global_filters.append(filter_func)

    def remove_global_filter(self, filter_func: EventFilter) -> bool:
        """
        Remove a global event filter.

        Args:
            filter_func: Filter function to remove.

        Returns:
            True if removed, False otherwise.
        """
        try:
            self._global_filters.remove(filter_func)
            return True
        except ValueError:
            return False

    def get_statistics(self) -> EventStatistics:
        """
        Get event statistics.

        Returns:
            Current event statistics.
        """
        return EventStatistics(
            total_events=self._stats.total_events,
            events_by_category=self._stats.events_by_category.copy(),
            events_by_severity=self._stats.events_by_severity.copy(),
            events_per_hour=self._calculate_events_per_hour(),
            last_event_time=self._stats.last_event_time,
        )

    def get_history(self, limit: int | None = None) -> list[BaseEvent]:
        """
        Get event history.

        Args:
            limit: Maximum number of events to return.

        Returns:
            List of recent events.
        """
        if limit is None:
            return list(self._event_history)
        return list(self._event_history)[-limit:]

    def get_queue_size(self) -> int:
        """Get the size of the event history."""
        return len(self._event_history)

    def get_subscriber_count(self) -> int:
        """Get the total number of subscribers."""
        count = 0
        for subscribers in self._subscribers.values():
            count += len(subscribers)
        return count

    def clear_history(self) -> None:
        """Clear event history."""
        self._event_history.clear()

    def clear_all_subscriptions(self) -> None:
        """Clear all subscriptions."""
        self._subscribers.clear()
        self._listeners.clear()

    def get_stats(self) -> dict[str, Any]:
        """
        Get detailed statistics.

        Returns:
            Dictionary with detailed statistics.
        """
        return {
            "total_events": self._stats.total_events,
            "last_event_time": self._stats.last_event_time,
            "events_by_category": self._stats.events_by_category,
            "events_by_severity": self._stats.events_by_severity,
            "events_per_hour": self._stats.events_per_hour,
            "subscriber_count": self.get_subscriber_count(),
            "transformer_count": sum(len(t) for t in self._transformers.values()),
            "global_filter_count": len(self._global_filters),
            "history_size": len(self._event_history),
        }

    async def set_use_queue_manager(self, enabled: bool) -> None:
        """
        Enable or disable queue manager integration.

        Args:
            enabled: Whether to use queue manager.
        """
        self._use_queue_manager = enabled
        if enabled and self._queue_manager is None:
            await self._setup_queue_manager_integration()

    def get_queue_manager(self) -> Any:
        """Get the queue manager instance."""
        return self._queue_manager

    # Private methods

    def _generate_subscriber_id(self) -> str:
        """Generate a unique subscriber ID."""
        return f"sub-{uuid4().hex[:12]}"

    def _update_statistics(self, event: BaseEvent) -> None:
        """Update event statistics."""
        self._stats.total_events += 1
        self._stats.last_event_time = datetime.utcnow()

        # Category count
        category = event.category
        if isinstance(category, EventCategory):
            category = category.value
        self._stats.events_by_category[category] = (
            self._stats.events_by_category.get(category, 0) + 1
        )

        # Severity count
        severity = event.severity
        if isinstance(severity, EventSeverity):
            severity = severity.value
        self._stats.events_by_severity[severity] = (
            self._stats.events_by_severity.get(severity, 0) + 1
        )

        # Hourly count
        hour = datetime.utcnow().hour
        self._stats.events_per_hour[hour] += 1

    def _apply_global_filters(self, event: BaseEvent) -> bool:
        """Apply global filters to event."""
        for filter_func in self._global_filters:
            if not filter_func(event):
                return False
        return True

    async def _apply_transformers(self, event: BaseEvent) -> BaseEvent:
        """Apply transformers to event."""
        transformed = event

        # Exact type match transformers
        if event.type in self._transformers:
            for transformer in self._transformers[event.type]:
                result = transformer(transformed)
                if asyncio.iscoroutine(result):
                    transformed = await result
                else:
                    transformed = result

        # Regex pattern transformers
        for pattern, transformers in self._transformers.items():
            if pattern != event.type:
                try:
                    regex = re.compile(pattern)
                    if regex.match(event.type):
                        for transformer in transformers:
                            result = transformer(transformed)
                            if asyncio.iscoroutine(result):
                                transformed = await result
                            else:
                                transformed = result
                except re.error:
                    pass

        return transformed

    async def _process_event_subscribers(self, event: BaseEvent) -> None:
        """Process all matching subscribers for an event."""
        all_subscribers: list[EventSubscriber] = []

        # Find all matching subscribers
        for key, subscribers in self._subscribers.items():
            for sub in subscribers:
                if sub.matches(event.type):
                    all_subscribers.append(sub)

        # Sort by priority (higher first)
        all_subscribers.sort(key=lambda s: s.options.priority, reverse=True)

        # Execute handlers
        for subscriber in all_subscribers:
            try:
                # Apply subscriber filter
                if subscriber.options.filter and not subscriber.options.filter(event):
                    continue

                # Execute handler
                result = subscriber.handler(event)
                if asyncio.iscoroutine(result):
                    await result

                # Remove if once
                if subscriber.options.once:
                    self.unsubscribe(subscriber.id)

            except Exception as e:
                # Create error event
                error_event = BaseEvent(
                    type="system:error",
                    category=EventCategory.SYSTEM,
                    severity=EventSeverity.ERROR,
                    source="EventEngine",
                    data={
                        "original_event_id": event.id,
                        "subscriber_id": subscriber.id,
                        "error": str(e),
                    },
                )
                await self._emit("event:error", error_event)

    async def _emit(self, event_type: str, event: BaseEvent) -> None:
        """Emit an internal event."""
        if event_type in self._listeners:
            for handler in self._listeners[event_type]:
                try:
                    result = handler(event)
                    if asyncio.iscoroutine(result):
                        await result
                except Exception:
                    pass  # Silently ignore internal event errors

    def _calculate_events_per_hour(self) -> float:
        """Calculate average events per hour."""
        total = sum(self._stats.events_per_hour)
        return total / 24.0

    async def _setup_queue_manager_integration(self) -> None:
        """Set up queue manager integration."""
        try:
            from .queue_manager import get_queue_manager

            self._queue_manager = get_queue_manager()

            # Register default processors
            self._queue_manager.register_processor(
                "default",
                self._process_queued_events,
            )
            self._queue_manager.register_processor(
                "priority",
                self._process_queued_events,
            )
            self._queue_manager.register_processor(
                "batch",
                self._process_batch_events,
            )
            self._queue_manager.register_processor(
                "failed",
                self._process_failed_events,
            )

        except ImportError:
            self._use_queue_manager = False

    async def _process_queued_events(self, events: list[BaseEvent]) -> None:
        """Process events from queue."""
        for event in events:
            await self._process_event_subscribers(event)

    async def _process_batch_events(self, events: list[BaseEvent]) -> None:
        """Process batch events."""
        # Group by type for batch processing
        events_by_type: dict[str, list[BaseEvent]] = {}
        for event in events:
            if event.type not in events_by_type:
                events_by_type[event.type] = []
            events_by_type[event.type].append(event)

        # Process each type group
        for event_list in events_by_type.values():
            for event in event_list:
                await self._process_event_subscribers(event)

    async def _process_failed_events(self, events: list[BaseEvent]) -> None:
        """Process failed events with retry logic."""
        for event in events:
            try:
                await self._process_event_subscribers(event)
            except Exception as e:
                # Log error but don't re-raise
                print(f"Failed to reprocess event {event.id}: {e}")


# Singleton accessor function
def get_event_engine() -> EventEngine:
    """
    Get the singleton EventEngine instance.

    Returns:
        EventEngine singleton instance.
    """
    return EventEngine.get_instance()


# Default singleton instance
event_engine = EventEngine.get_instance()
